jupyterhub/apihandlers/base.py

```diff
@@ -79,23 +79,32 @@ def get_scope_filter(self, req_scope):
                 % req_scope,
             )
 
-        def has_access(orm_resource, kind):
+        def has_access_to(orm_resource, kind):
             """
-            param orm_resource: User or Service or Group
-            param kind: 'users' or 'services' or 'groups'
+            param orm_resource: User or Service or Group or spawner
+            param kind: 'user' or 'service' or 'group' or 'server'.
+            `kind` could probably be derived from `orm_resource`, problem is Jupyterhub.users.User
             """
             if sub_scope == scopes.Scope.ALL:
                 return True
             else:
                 found_resource = orm_resource.name in sub_scope[kind]
             if not found_resource:  # Try group-based access
-                if 'group' in sub_scope and kind == 'user':
+                if kind == 'server':
+                    # First check if we have access to user info
+                    user_name = orm_resource.user.name
+                    found_resource = user_name in sub_scope['user']
+                    if not found_resource:
+                        # Now check for specific servers:
+                        server_format = f"{orm_resource.user / orm_resource.name}"
+                        found_resource = server_format in sub_scope[kind]
+                elif 'group' in sub_scope and kind == 'user':
```

Comment:
What happens for `kind='server'` when access is granted through a group? Can we make sure to have a test for that?